### PR TITLE
Simple expansions in stream name, and arbitrary codec support.

### PR DESCRIPTION
**Pull request by samcday**
The following tokens, if supplied in stream name, are expanded:
- `%hostname%` to the FQDN of the local machine.
- `%ipv4%` to the first non-local IP in the interface list
- `%instance_id%` to the local EC2 instance id.

The plugin is now also able to use codecs, instead of just writing the event's
"message" field. An obvious use case for this is writing JSON to CW Logs:

Altogether, now:

``` nginx
cloudwatchlogs {
    region => "foo"
    codec => json {}
    log_group_name => "bacon"
    log_stream_name => "%instance_id%"
    use_codec => true
}
```

### CODE REVIEW

File: lib/logstash/outputs/cloudwatchlogs.rb [hunk 1]
```diff
@@ -219,7 +246,7 @@ def put_log_events(log_events)
 
   private
   def invalid?(event)
-    status = event[TIMESTAMP].nil? || event[MESSAGE].nil?
+    status = event[TIMESTAMP].nil? || (!@use_codec && event[MESSAGE].nil?)
     if status
```
Comment (jtblin) [inline lib/logstash/outputs/cloudwatchlogs.rb:250]: Nit: `@logger.warn("Skipping invalid event #{event.to_hash}") if status` is more idiomatic.

**Comment by johnjeffers** [comment]
+1! Could really use the codec functionality today, for streaming into Cloudwatch Logs, as mentioned above.

**Comment by samcday** [comment]
Hey @wanghq - can we get this one reviewed and merged plx? :)

**Comment by johnjeffers** [comment]
Still waiting for this to be merged, please.

**Comment by johnjeffers** [comment]
It's been 5 months! Can we get this merged???

**Comment by samcday** [comment]
I suspect this repo is dead and no longer maintained :(

**Comment by rmw-at-amazon** [comment]
Sorry for the delay in accepting the pull request, it's not dead and it's maintained, but we just missed it.  Apologies and love the change.

**Pull request merged**